crates/re_types_builder/src/lib.rs

```diff
@@ -431,22 +524,133 @@ pub fn generate_python_code(
+        // If the prefix is pkg_path, strip it, and then append to tempdir
+        // However, if the prefix is testing_pkg_path, strip it and insert an extra
+        // "testing" to avoid name collisions.
+        filepath.strip_prefix(pkg_path).map_or_else(
+            |_| {
+                tempdir_path
+                    .join("testing")
+                    .join(filepath.strip_prefix(testing_pkg_path).unwrap())
+            },
+            |f| tempdir_path.join(f),
+        )
+    }
+
+    fn format_python_dir(dir: &Utf8PathBuf) -> anyhow::Result<()> {
+        re_tracing::profile_function!();
+
+        // The order below is important and sadly we need to call black twice. Ruff does not yet
+        // fix line-length (See: https://github.com/astral-sh/ruff/issues/1904).
+        //
+        // 1) Call black, which among others things fixes line-length
+        // 2) Call ruff, which requires line-lengths to be correct
+        // 3) Call black again to cleanup some whitespace issues ruff might introduce
+
+        run_black_on_dir(dir).context("black")?;
+        run_ruff_on_dir(dir).context("ruff")?;
+        run_black_on_dir(dir).context("black")?;
+        Ok(())
+    }
+
+    fn python_project_path() -> Utf8PathBuf {
+        let path = crate::rerun_workspace_path()
+            .join("rerun_py")
+            .join("pyproject.toml");
+        assert!(path.exists(), "Failed to find {path:?}");
+        path
+    }
+
+    fn run_black_on_dir(dir: &Utf8PathBuf) -> anyhow::Result<()> {
+        re_tracing::profile_function!();
+        use std::process::{Command, Stdio};
+
+        let proc = Command::new("black")
+            .arg(format!("--config={}", python_project_path()))
+            .arg(dir)
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped())
+            .spawn()?;
+
+        let output = proc.wait_with_output()?;
+
+        if output.status.success() {
+            Ok(())
+        } else {
+            let stdout = String::from_utf8_lossy(&output.stdout).trim().to_owned();
+            let stderr = String::from_utf8_lossy(&output.stderr).trim().to_owned();
+            anyhow::bail!("{stdout}\n{stderr}")
+        }
+    }
+
+    fn run_ruff_on_dir(dir: &Utf8PathBuf) -> anyhow::Result<()> {
+        re_tracing::profile_function!();
+        use std::process::{Command, Stdio};
+
+        let proc = Command::new("ruff")
+            .arg(format!("--config={}", python_project_path()))
+            .arg("--fix")
+            .arg(dir)
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped())
+            .spawn()?;
+
+        let output = proc.wait_with_output()?;
+
+        if output.status.success() {
+            Ok(())
+        } else {
+            let stdout = String::from_utf8_lossy(&output.stdout).trim().to_owned();
+            let stderr = String::from_utf8_lossy(&output.stderr).trim().to_owned();
+            anyhow::bail!("{stdout}\n{stderr}")
+        }
+    }
```

Comment:
I don't understand why you moved out the language-specific formatting out of the language-specific modules?

To me it makes more sense if `GeneratedFiles` would be the final files, with no extra processing required.
Then these functions would be a lot more similar to each other (identical?)

Reply:
…and it would make the PR diff smaller too :)

Reply:
I just don't like the the mixup of concerns: the codegen pass should only concern itself with generating code, not any post-processing (including formatting) nor I/O.
Especially so with the current implementation where the codegen pass is nice and pure while the formatting pass requires I/O, process spawning, etc.

I will likely introduce a formalized formatting pass in https://github.com/rerun-io/rerun/issues/3495 though, which will make all these `generate_XXX` methods go away and make `--check`ing trivial.

Reply:
that makes sense!